src/resources.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Structs to manage device resources.
+
+/// Enumeration for device resources.
+#[allow(missing_docs)]
+#[derive(Clone)]
+pub enum DeviceResourceEntry {
+    /// IO Port address range.
+    PioAddressRange { base: u16, size: u16 },
+    /// Memory Mapped IO address range.
+    MmioAddressRange { base: u64, size: u64 },
+    /// Legacy IRQ number.
+    LegacyIrq(u32),
+    /// PCI MSI IRQ numbers.
+    PciMsiIrq { base: u32, size: u32 },
+    /// PCI MSIx IRQ numbers.
+    PciMsixIrq { base: u32, size: u32 },
+    /// Generic MSI IRQ numbers.
+    GenericIrq { base: u32, size: u32 },
+    /// Network Interface Card MAC address.
+    NicMacAddresss(String),
+    /// KVM memslot index.
+    KvmMemSlot(u32),
+}
+
+/// Newtype to store a set of device resources.
+#[derive(Default, Clone)]
+pub struct DeviceResources(Vec<DeviceResourceEntry>);
+
+impl DeviceResources {
+    /// Create a container object to store device resources.
+    pub fn new() -> Self {
+        DeviceResources(Vec::new())
+    }
+
+    /// Append a device resource to the container object.
+    pub fn append(&mut self, entry: DeviceResourceEntry) {
+        self.0.push(entry);
+    }
+
+    /// Get the IO port address resources.
+    pub fn get_pio_address_ranges(&self) -> Vec<(u16, u16)> {
+        let mut vec = Vec::new();
+        for entry in self.0.iter().as_ref() {
+            if let DeviceResourceEntry::PioAddressRange { base, size } = entry {
+                vec.push((*base, *size));
+            }
+        }
+        vec
+    }
+
+    /// Get the Memory Mapped IO address resources.
+    pub fn get_mmio_address_ranges(&self) -> Vec<(u64, u64)> {
+        let mut vec = Vec::new();
+        for entry in self.0.iter().as_ref() {
+            if let DeviceResourceEntry::MmioAddressRange { base, size } = entry {
+                vec.push((*base, *size));
+            }
+        }
+        vec
+    }
+
+    /// Get the legacy interrupt number(IRQ).
+    pub fn get_legacy_irq(&self) -> Option<u32> {
+        for entry in self.0.iter().as_ref() {
+            if let DeviceResourceEntry::LegacyIrq(base) = entry {
```

Comment:
Here (and below several routines) you seem to assume there is only one irq entry existed, which if it is true, then you probably need check that in append() above.